### ISSUE TITLE
[RFC] Should the upload handler be uninstallable?

### ISSUE THREAD

**Issue by Maikuolan** [Maikuolan/phpMussel#218]
### __Should the upload handler be uninstallable?__

(Currently, it isn't).

__Any backwards-compatibility concerns?__

Nope. :-)

Because the loader checks for the existence of the upload handler before attempting to call it, its non-existence shouldn't create any errors/warnings/etc (excerpt from `loader.php`).

```PHP
    /**
     * Check whether the upload handler exists and attempt to load it.
     */
    if (file_exists($phpMussel['Vault'] . 'upload.php') && !$phpMussel['Alternate']) {
        require $phpMussel['Vault'] . 'upload.php';
    }
```

Its non-existence doesn't result in any alternative action. Rather, the loader will just continue along on its merry way until the entire execution finishes as per normal.

__So, how much work would be involved to achieve this?__

Would literally need to add just one line only to the `components.dat` file:

```diff
 phpMussel Upload Handler:
  Name: phpMussel Upload Handler
  Version: 2.1.3-DEV+20136416
  Minimum Required: 2.0.0
  Changelog: https://raw.githubusercontent.com/phpMussel/phpMussel/v2/Changelog-v2.txt
  Remote: https://raw.githubusercontent.com/phpMussel/phpMussel/v2/vault/components.dat
+ Uninstallable: true
  Tests: https://api.github.com/repos/phpMussel/phpMussel/commits/v2/status
  Files:
   From:
    - https://raw.githubusercontent.com/phpMussel/phpMussel/v2/vault/template_custom.html
    - https://raw.githubusercontent.com/phpMussel/phpMussel/v2/vault/template_default.html
    - https://raw.githubusercontent.com/phpMussel/phpMussel/v2/vault/upload.php
   To:
    - template_custom.html
    - template_default.html
    - upload.php
   Checksum:
    - 68d157607aed65d23212d5e81b939ef9529d532b7bb679eb207a37604d444143:1065
    - b50e0fed13dfc51fffcb2ef6a657f9977220ab781c688c34000ac2503d0c5c1b:2031
    - 4daf7be09f676dc25ac276cff82036ebbc692d633053f3d1c03c44638dfecefd:21263
  Reannotate: components.dat
```

..that's it. :-)

Once done, it would become possible to entirely remove the upload handler simply by clicking to uninstall it via the front-end updates page.

However, worth nothing that this would make it uninstallable only for phpMussel v2 and onward, because whereas the upload handler and phpMussel core are regarded as two separate components by phpMussel v2's components metadata, they aren't by phpMussel v1's metadata (the metadata for the uploader handler is bundled in with that of the core in v1). So, not a backwards-compatible change (seeing as v2 is the current latest major version), but also not a v1-compatible change.

The actual files removed when uninstalling the upload handler, as per the excerpt above, would be `template_custom.html`, `template_default.html`, and `upload.php`.

__Why? What scenarios or use-cases are there where this might be desirable?__

Having a way to disable the upload handler would be useful for those which don't actually need automated upload scanning, but instead want to utilise phpMussel from within an already existing framework/system/etc, whereby automated upload scanning could potentially actually get in the way.

Pull request [#182](https://github.com/phpMussel/phpMussel/pull/182) has also been sitting there, remaining open but unmerged, for what is now just a little shy of two years, basically because nobody (myself in particular) has been 100% sure about the best way to approach the problem which that pull request was intended to address. We've already got "maintenance mode", but that's potentially a little ham-fisted for some use-cases, and also isn't entirely succinct about a user's true intentions. We could add additional configuration options/directives, but then we'll have multiple options/directives that basically do the same thing, which isn't particularly elegant. So, what else to do..?

I could be lazy and just tell people to manually delete the files, but that's also not particularly elegant, and also isn't a solution compatible with the mechanisms utilised by the front-end updates page, because attempting to update or repair components will likely result in the deleted files being automatically recreated anyway. Just telling people to uninstall it also feels just slightly lazy and inelegant too, in a way, but doesn't feel quite as much so as it does with telling people to manually delete files, and uninstalling is also the best way to ensure the files actually stay deleted for those users which utilise the front-end updates page.

It should be noted that this is *not* an attempt to dismiss the idea of eventually having separate repositories for the core, front-end, uploader handler, etc, in order to provide a more Composer-friendly version of phpMussel at some point in the future. Rather, I'm just thinking it could serve as a pathway to deal with the problem with what we've got right now, until the separate repositories idea eventuates, and maybe as an excuse to call the aforementioned pull request resolved (because the question of a dedicated configuration option won't be lingering anymore).

Also, this particular point mentioned in the aforementioned pull request:
> The auto-scanning and the frontend for updating actually get a little bit in the way for the scenario I'm considering, as it would seem more appropriate to deploy virus definition files from version control as part of the app, rather than individually on each server through the frontend.

..has actually come up again recently (with some permutation), at our Gitter channel:
> I found that this was also the only way I could call the phpmussel routines from a web page without having the page redirect to the inflexible PHP Mussel 'Virus upload detected' screen, and be able to control how I handled the message display.

..and also once from a user that I've been communicating with over Facebook.

So, in addition to the main question here, I would also ask:
- Do you think this would be sufficient enough to resolve the aforementioned use-cases and scenarios (at least, for the moment, until something better, like possibly the multiple repositories idea, eventuates)?
- Could this solution (just uninstalling it) be regarded as "good advice" for such use-cases?

A possible extenuating problem: The `components.dat` file, as well as the files in question to be uninstallable, as well as phpMussel's default configuration file, are still a part of this main repository here, which means that they'll likely all be reverted/recreated/reinstalled when updating via Composer. I'll figure something out for the configuration problem a little later, and the problem of files being recreated should eventually resolve itself once we have separate repositories for everything in the future anyway. But, I mention this here anyway, just because it might have some bearing on opinions regarding the actual questions I'm asking here.

(Completely different and alternative ideas and solutions are also welcome).

Cheers. :-)

**Comment by Maikuolan** [comment]
> Would literally need to add just one line only to the `components.dat` file:

..make that two lines. Forgot about a very small safety precaution I'd added into the main front-end file a while back (excerpt from `frontend.php`, attached below). Remembered just now, after creating this issue. My bad. '^.^

```PHP
) . '|phpMussel.*|Common Classes Package)$~i', $phpMussel['Components']['Key']) || $phpMussel['IsInUse'](
```

Anyway, still an extremely small amount of work involved. Should just be a matter of changing that `|phpMussel.*|` to `|phpMussel(?! Upload handler).*|` or something like that (not tested yet, but anyway, it's just RFC right now, so I can figure that out later if it seems like a good idea).

**Comment by Maikuolan** [comment]
Implemented (c9e582871a3a39bf705edd8eee910b65e6b0ae44). Closing.
